File: CP/LeetCode/Python/219.py
# 219. Contains Duplicate II

from __future__ import annotations


class Solution:
    def containsNearbyDuplicate(self, nums: list[int], k: int) -> bool:
        indexes = {}
        for index, num in enumerate(nums):
            if num in indexes and abs(index - indexes[num]) <= k:
                return True

            indexes[num] = index
        return False


if __name__ == "__main__":
    s = Solution()
    contains = s.containsNearbyDuplicate([1, 2, 3, 1], 3)

    print(contains)

File: CP/LeetCode/Java/219.java
// 219. Contains Duplicate II

import java.util.HashMap;
import java.util.Map;

class Solution {
    public boolean containsNearbyDuplicate(final int[] nums, final int k) {
        Map<Integer, Integer> map = new HashMap<>();
        int n = nums.length;

        for (int i = 0; i < n; i++) {
            if (map.containsKey(nums[i]) && i - map.get(nums[i]) <= k) {
                return true;
            }
            map.put(nums[i], i);
        }
        return false;
    }

    public static void main(final String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.containsNearbyDuplicate(new int[] { 1, 2, 3, 1 }, 3));
        System.out.println(solution.containsNearbyDuplicate(new int[] { 1, 0, 1, 1 }, 1));
        System.out.println(solution.containsNearbyDuplicate(new int[] { 1, 2, 3, 1, 2, 3 }, 2));
    }
}
